### PR TITLE
Add horovod estimator save/load/get_models/shutdown

### PR DESCRIPTION
**Pull request by yushan111**

### CODE REVIEW

File: pyzoo/zoo/orca/learn/pytorch/torch_runner.py [hunk 1]
```diff
@@ -186,8 +186,7 @@ def train_epoch(self,
             SCHEDULER_STEP: self.scheduler_step_freq
         })
         with self.timers.record("train_epoch"):
-            if TorchRunner.should_wrap_dataloader(data_loader):
```
Comment (hkvision) [inline pyzoo/zoo/orca/learn/pytorch/torch_runner.py:189]: Why change this?

File: pyzoo/zoo/orca/learn/pytorch/pytorch_horovod_estimator.py [hunk 2]
```diff
@@ -315,3 +314,87 @@ def validate(self, data_creator, num_steps=None, profile=False, info=None):
...
+    def save(self, checkpoint):
+        """Saves the Estimator state to the provided checkpoint path.
+
+        :param checkpoint: (str) Path to target checkpoint file.
+        """
+        state_dict = self.save_state_dict()
+        torch.save(state_dict, checkpoint)
+        return checkpoint
+
+    def save_state_dict(self):
```
Comment (jason-dai) [inline pyzoo/zoo/orca/learn/pytorch/pytorch_horovod_estimator.py:342]: get_state_dict

File: pyzoo/zoo/orca/learn/pytorch/pytorch_horovod_estimator.py [hunk 3]
```diff
@@ -315,3 +314,87 @@ def validate(self, data_creator, num_steps=None, profile=False, info=None):
...
+        models = self.model_creator(self.config)
+        if not isinstance(models, collections.Iterable):
+            models = [models]
+
+        for model, state_dict in zip(models, state["models"]):
+            model.load_state_dict(state_dict)
+            unwrapped += [model.module if hasattr(model, "module") else model]
+        if len(unwrapped) == 1:
+            return unwrapped[0]
+        return unwrapped
```
Comment (jason-dai) [inline pyzoo/zoo/orca/learn/pytorch/pytorch_horovod_estimator.py:331]: I think you should always return just one model, no?

File: pyzoo/zoo/orca/learn/pytorch/torch_runner.py (hunk 1 above)
Comment (yushan111) [reply pyzoo/zoo/orca/learn/pytorch/torch_runner.py:189]: if `should_wrap_dataloader`, then `with_sampler` (the data loader wrapper) should be called. `iter(data_loader)` is not relevant to `should_wrap_dataloader`.

File: pyzoo/zoo/orca/learn/pytorch/pytorch_horovod_estimator.py (hunk 3 above)
Comment (yushan111) [reply pyzoo/zoo/orca/learn/pytorch/pytorch_horovod_estimator.py:331]: yes, we only support one model and one optimizer for now. Do we also need to simplify [convert to models]( https://github.com/intel-analytics/analytics-zoo/blob/master/pyzoo/zoo/orca/learn/pytorch/pytorch_horovod_estimator.py#L45-L46) ?

File: pyzoo/zoo/orca/learn/pytorch/pytorch_horovod_estimator.py [hunk 4]
```diff
@@ -315,3 +314,80 @@ def validate(self, data_creator, num_steps=None, profile=False, info=None):
...
+        state_dict = self.get_state_dict()
+        torch.save(state_dict, checkpoint)
+        return checkpoint
+
+    def get_state_dict(self):
+        stream_ids = [
+            worker.state_stream.remote()
+            for worker in self.remote_workers
+        ]
+        [stream_id], stream_ids = ray.wait(stream_ids)
```
Comment (yangw1234) [inline pyzoo/zoo/orca/learn/pytorch/pytorch_horovod_estimator.py:340]: add a comment to explain what this means?

File: pyzoo/zoo/orca/learn/pytorch/pytorch_horovod_estimator.py [hunk 5]
```diff
@@ -315,3 +314,80 @@ def validate(self, data_creator, num_steps=None, profile=False, info=None):
...
+
+    def load(self, checkpoint):
+        """Loads the Estimator and all workers from the provided checkpoint.
+
+        :param checkpoint: (str) Path to target checkpoint file.
+        """
+        state_dict = torch.load(checkpoint)
+        self.load_state_dict(state_dict)
+
+    def load_state_dict(self, state_dict, blocking=False):
```
Comment (yangw1234) [inline pyzoo/zoo/orca/learn/pytorch/pytorch_horovod_estimator.py:356]: how about setting blocking to True by default?

**Comment by yushan111** [comment]
Tests:

- local

  - [pytorch estimator example](https://github.com/intel-analytics/analytics-zoo/blob/b81a6e23d528af85e9e2907dd4ce35b678323462/pyzoo/zoo/examples/orca/learn/horovod/pytorch_estimator.py)
  - [UT](https://github.com/intel-analytics/analytics-zoo/blob/b81a6e23d528af85e9e2907dd4ce35b678323462/pyzoo/test/zoo/orca/learn/ray/pytorch/test_pytorch_estimator.py) 

- distributed
  - [pytorch estimator example](https://github.com/intel-analytics/analytics-zoo/blob/b81a6e23d528af85e9e2907dd4ce35b678323462/pyzoo/zoo/examples/orca/learn/horovod/pytorch_estimator.py)
  - [test_save_and_restore](https://github.com/intel-analytics/analytics-zoo/blob/b81a6e23d528af85e9e2907dd4ce35b678323462/pyzoo/test/zoo/orca/learn/ray/pytorch/test_pytorch_estimator.py#L55)
- jenkins
  http://10.239.47.210:18888/view/ZOO-PR/job/ZOO-PR-Validation/3987/

File: pyzoo/zoo/orca/learn/pytorch/pytorch_horovod_estimator.py (hunk 4 above)
Comment (yushan111) [reply pyzoo/zoo/orca/learn/pytorch/pytorch_horovod_estimator.py]: Done

File: pyzoo/zoo/orca/learn/pytorch/pytorch_horovod_estimator.py (hunk 5 above)
Comment (yushan111) [reply pyzoo/zoo/orca/learn/pytorch/pytorch_horovod_estimator.py]: Done

**Comment by yushan111** [comment]
Merge for now since tests passed.

**Pull request merged**